### PR TITLE
Update the subscription models to align on CAMARA commonalities

### LINKED ISSUE #162: Align subscription APIs with new template
**Problem description**
The newly created subscription APIs (roaming and reachability) need to be updated to comply with the [subscription template](https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml)

**Possible evolution**
Create a PR to adapt both APIs. No additional changes should be made.

**Alternative solution**

**Additional context**

### PR DESCRIPTION
**Pull request by maxl2287**
#### What type of PR is this?

Add one of the following kinds:

* correction
* enhancement/feature

#### What this PR does / why we need it:
Update of the subscription-models based on: 
- https://github.com/camaraproject/Commonalities/blob/main/documentation/API-design-guidelines.md#12-subscription-notification--event
- https://github.com/camaraproject/Commonalities/blob/main/artifacts/camara-cloudevents/event-subscription-template.yaml

#### Which issue(s) this PR fixes:

Fixes #162

### CODE REVIEW

**Comment by akoshunyadi** [comment]
The base url should be with apiName, so 'device-' is missing..  (I can't comment there directly)

File: code/API_definitions/device-roaming-status-subscriptions.yaml
Comment (akoshunyadi) [inline code/API_definitions/device-roaming-status-subscriptions.yaml:704]: I guess we need here device- too, and then in the scopes too.

**Review by sachinvodafone — APPROVED** [review]

File: code/API_definitions/device-roaming-status-subscriptions.yaml
Comment (maxl2287) [reply code/API_definitions/device-roaming-status-subscriptions.yaml:704]: Aggree 🚀

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (akoshunyadi) [inline code/API_definitions/device-reachability-status-subscriptions.yaml:39]: Rather "It is used in following cases" to emphasize that those lines are OR connected

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (akoshunyadi) [inline code/API_definitions/device-reachability-status-subscriptions.yaml:62]: This block seems to be duplicated

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (akoshunyadi) [inline code/API_definitions/device-reachability-status-subscriptions.yaml:135]: we might want to rename connectivity to reachability here too

File: code/API_definitions/device-roaming-status-subscriptions.yaml
Comment (akoshunyadi) [inline code/API_definitions/device-roaming-status-subscriptions.yaml:41]: Rather "It is used in following cases" to emphasize that those lines are OR connected

File: code/API_definitions/device-roaming-status-subscriptions.yaml
Comment (akoshunyadi) [inline code/API_definitions/device-roaming-status-subscriptions.yaml:56]: The block "Authentication..." seems to be missing here

**Review by bigludo7 — DISMISSED** [review]
LGTM
Nice job Max ! thanks

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (sachinvodafone) [inline code/API_definitions/device-reachability-status-subscriptions.yaml:35]: It stated "reachability disconnected" but didn't specify whether it pertains to "SMS" or "data," leading us to assume it applies to both. So just confirming if there is no use-case to send the notification separately for each.

**Comment by sachinvodafone** [comment]
As per today discussion over subscription response, would apprecaite if we can include examples in specification that pointed in [discussion](https://github.com/camaraproject/DeviceStatus/discussions/168)

**Comment by bigludo7** [comment]
@maxl2287 I have a second thought about making the device identifier optional in the subscription.
My point is the following: 
For the POST that's fine to not have this device identifier as we have it thru the access token.

But, when the developer check the subscription 10 days later, in the GET /subscriptions response better to have this device identifier present because without it impossible to know what device is targeted....

So... either 1/ we keep the device mandatory or 2/we make it optionnal for the POST request but mandatory in the GET response

I prefer 1/ but perhaps we can have other solution.

cc: @fernandopradocabrillo

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (maxl2287) [reply code/API_definitions/device-reachability-status-subscriptions.yaml:35]: As I understand it from the current specification, this `reachability-disconnected` - events occure when the device is not connected at all.
So you want to be informed when there is neither DATA nor SMS connection.

Please correct me if I am wrong @bigludo7

**Comment by maxl2287** [comment]
> @maxl2287 I have a second thought about making the device identifier optional in the subscription. My point is the following: For the POST that's fine to not have this device identifier as we have it thru the access token.
> 
> But, when the developer check the subscription 10 days later, in the GET /subscriptions response better to have this device identifier present because without it impossible to know what device is targeted....
> 
> So... either 1/ we keep the device mandatory or 2/we make it optionnal for the POST request but mandatory in the GET response
> 
> I prefer 1/ but perhaps we can have other solution.
> 
> cc: @fernandopradocabrillo

Are you talking about https://github.com/camaraproject/DeviceStatus/issues/173 ?

**Comment by maxl2287** [comment]
> As per today discussion over subscription response, would apprecaite if we can include examples in specification that pointed in [discussion](https://github.com/camaraproject/DeviceStatus/discussions/168)

@sachinvodafone I have added three examples for GET /subscriptions/{subscriptionId} pointing to specific `status` of the subscription.

**Review by sachinvodafone — DISMISSED** [review]

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (sachinvodafone) [reply code/API_definitions/device-reachability-status-subscriptions.yaml:35]: I was reconfirming  if there is a separate event triggered when the device connects to the network for SMS and Data individually. Why do we only have one event triggered for disconnection?

**Comment by sachinvodafone** [comment]
> > As per today discussion over subscription response, would apprecaite if we can include examples in specification that pointed in [discussion](https://github.com/camaraproject/DeviceStatus/discussions/168)
> 
> @sachinvodafone I have added three examples for GET /subscriptions/{subscriptionId} pointing to specific `status` of the subscription.

Thanks @maxl2287 , Can we get an example for "GET /subscriptions" where we provide only the subscriptionID for subscriptions with incomplete asynchronous calls and show the full response for the others (synchronous)?

**Comment by maxl2287** [comment]
> Thanks @maxl2287 , Can we get an example for "GET /subscriptions" where we provide only the subscriptionID for subscriptions with incomplete asynchronous calls and show the full response for the others (synchronous)?

@sachinvodafone 
However, this will not align with the specified response for GET `/subscriptions`. 
The response for this endpoint includes a list of only having `Subscription` components. 

There should not be a mix of `Subscription` objects and `subscriptionId` values within the same list.

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (bigludo7) [reply code/API_definitions/device-reachability-status-subscriptions.yaml:35]: @maxl2287 Indeed as of now reachability-disconnected is when the mobile is not connected at all.

@sachinvodafone I got your point.
We have today 3 notification types: Data Connected, SMS connected and Disconnected. SMS connected mean that data is disconnected.
We have aligned this on the synchronous APIs response.

Alternatively we can define slightly distinct event as
- data-on 
- data-off
- sms-on
- sms-off

Max, Sachin wdyt?

**Comment by bigludo7** [comment]
> > @maxl2287 I have a second thought about making the device identifier optional in the subscription. My point is the following: For the POST that's fine to not have this device identifier as we have it thru the access token.
> > But, when the developer check the subscription 10 days later, in the GET /subscriptions response better to have this device identifier present because without it impossible to know what device is targeted....
> > So... either 1/ we keep the device mandatory or 2/we make it optionnal for the POST request but mandatory in the GET response
> > I prefer 1/ but perhaps we can have other solution.
> > cc: @fernandopradocabrillo
> 
> @bigludo7 I aggree also on letting it mandatory here.

I suggest we do this way - let keep it mandatory. If in future we shift to optional this is not a breaking change.

**Comment by sachinvodafone** [comment]
> > Thanks @maxl2287 , Can we get an example for "GET /subscriptions" where we provide only the subscriptionID for subscriptions with incomplete asynchronous calls and show the full response for the others (synchronous)?
> 
> @sachinvodafone However, this will not align with the specified response for GET `/subscriptions`. The response for this endpoint includes a list of only having `Subscription` components.
> 
> There should not be a mix of `Subscription` objects and `subscriptionId` values within the same list.

@maxl2287 I understand your point that it will only include those subscriptions that have already been created. Thank you.

**Comment by maxl2287** [comment]
> > > Thanks @maxl2287 , Can we get an example for "GET /subscriptions" where we provide only the subscriptionID for subscriptions with incomplete asynchronous calls and show the full response for the others (synchronous)?
> > 
> > 
> > @sachinvodafone However, this will not align with the specified response for GET `/subscriptions`. The response for this endpoint includes a list of only having `Subscription` components.
> > There should not be a mix of `Subscription` objects and `subscriptionId` values within the same list.
> 
> @maxl2287 I understand your point that it will only include those subscriptions that have already been created. Thank you.

@sachinvodafone yes, but to be more precise: 
The list includes every subscription, which was created by the user, even those which are still in the creation process (see status ACTIVATION_REQUESTED).

**Comment by sachinvodafone** [comment]
> > > > Thanks @maxl2287 , Can we get an example for "GET /subscriptions" where we provide only the subscriptionID for subscriptions with incomplete asynchronous calls and show the full response for the others (synchronous)?
> > > 
> > > 
> > > @sachinvodafone However, this will not align with the specified response for GET `/subscriptions`. The response for this endpoint includes a list of only having `Subscription` components.
> > > There should not be a mix of `Subscription` objects and `subscriptionId` values within the same list.
> > 
> > 
> > @maxl2287 I understand your point that it will only include those subscriptions that have already been created. Thank you.
> 
> @sachinvodafone yes, but to be more precise: The list includes every subscription, which was created by the user, even those which are still in the creation process (see status ACTIVATION_REQUESTED).

Suppose if we have only one subscription and that is asynchronous which still not completed, in this case ,we will get "ACTIVATION_REQUESTED"  if user try for "GET /Subscriptions" . If Yes, then I am having another question.

**Comment by maxl2287** [comment]
> Suppose if we have only one subscription and that is asynchronous which still not completed, in this case ,we will get "ACTIVATION_REQUESTED" if user try for "GET /Subscriptions" . If Yes, then I am having another question.

You will receive something like:

```json
[
  {
    "sink": "https://endpoint.example.com/sink",
    "sinkCredential": {},
    "types": [
      "string"
    ],
    "config": {
      "subscriptionDetail": {
        "device": {
          "phoneNumber": "+123456789",
          "networkAccessIdentifier": "123456789@domain.com",
          "ipv4Address": {
            "publicAddress": "84.125.93.10",
            "publicPort": 59765
          },
          "ipv6Address": "2001:db8:85a3:8d3:1319:8a2e:370:7344"
        }
      },
      "subscriptionExpireTime": "2023-01-17T13:18:23.682Z",
      "subscriptionMaxEvents": 5,
      "initialEvent": true
    },
    "id": "550e8400-e29b-41d4-a716-446655440000",
    "startsAt": "2024-07-04T11:44:36.664Z",
    "expiresAt": "2024-07-04T11:44:36.664Z",
    "status": "ACTIVATION_REQUESTED"
  }
]
```

**Comment by sachinvodafone** [comment]
> > Suppose if we have only one subscription and that is asynchronous which still not completed, in this case ,we will get "ACTIVATION_REQUESTED" if user try for "GET /Subscriptions" . If Yes, then I am having another question.
> 
> You will receive something like:
> 
> ```json
> [
>   {
>     "sink": "https://endpoint.example.com/sink",
>     "sinkCredential": {},
>     "types": [
>       "string"
>     ],
>     "config": {
>       "subscriptionDetail": {
>         "device": {
>           "phoneNumber": "+123456789",
>           "networkAccessIdentifier": "123456789@domain.com",
>           "ipv4Address": {
>             "publicAddress": "84.125.93.10",
>             "publicPort": 59765
>           },
>           "ipv6Address": "2001:db8:85a3:8d3:1319:8a2e:370:7344"
>         }
>       },
>       "subscriptionExpireTime": "2023-01-17T13:18:23.682Z",
>       "subscriptionMaxEvents": 5,
>       "initialEvent": true
>     },
>     "id": "550e8400-e29b-41d4-a716-446655440000",
>     "startsAt": "2024-07-04T11:44:36.664Z",
>     "expiresAt": "2024-07-04T11:44:36.664Z",
>     "status": "ACTIVATION_REQUESTED"
>   }
> ]
> ```

I have updated the scenario in discussion board [here](https://github.com/camaraproject/DeviceStatus/discussions/168)

**Comment by maxl2287** [comment]
@bigludo7 can we do here a final review?
I just have added some examples based on @sachinvodafone's feedback.

**Review by bigludo7 — DISMISSED** [review]
LGTM.
Thanks

**Comment by bigludo7** [comment]
> @bigludo7 can we do here a final review? I just have added some examples based on @sachinvodafone's feedback.

Thanks @maxl2287 - Look good for me. Let's get final review from @sachinvodafone and @akoshunyadi and we're good to go.

**Review by akoshunyadi — COMMENTED** [review]
The server url seems to miss the device- prefix for roaming

**Review by akoshunyadi — DISMISSED** [review]

**Review by fernandopradocabrillo — APPROVED** [review]
LGTM

**Review by bigludo7 — DISMISSED** [review]
LGTM

**Comment by bigludo7** [comment]
@sachinvodafone it is ok for you now? Looking for you thumb up to merge :)
Thanks

File: code/API_definitions/device-reachability-status-subscriptions.yaml
Comment (maxl2287) [reply code/API_definitions/device-reachability-status-subscriptions.yaml:36]: @bigludo7 can we either create a discussion on that topic or an issue?
Maybe a better place to discuss :)

Comment (bigludo7) [reply code/API_definitions/device-reachability-status-subscriptions.yaml:36]: @maxl2287  Yes a discussion is a good idea. Do not block the PR for that. I copy/past las comment in a discussion.

**Review by bigludo7 — APPROVED** [review]
LGTM

**Review by akoshunyadi — APPROVED** [review]

**Pull request merged**
